plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -229,81 +228,79 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+                    if (!fileModifiedTimeDomain.includesNullableValue(packDateTimeWithZone(fileModifiedTime, UTC_KEY))) {
+                        continue;
+                    }
+                }
 
-            Schema fileSchema = scanTask.spec().schema();
-            Map<Integer, Optional<String>> partitionKeys = getPartitionKeys(scanTask);
+                Schema fileSchema = wholeFileTask.spec().schema();
+                Map<Integer, Optional<String>> partitionKeys = getPartitionKeys(wholeFileTask);
 
-            Set<IcebergColumnHandle> identityPartitionColumns = partitionKeys.keySet().stream()
-                    .map(fieldId -> getColumnHandle(fileSchema.findField(fieldId), typeManager))
-                    .collect(toImmutableSet());
+                Set<IcebergColumnHandle> identityPartitionColumns = partitionKeys.keySet().stream()
+                        .map(fieldId -> getColumnHandle(fileSchema.findField(fieldId), typeManager))
+                        .collect(toImmutableSet());
 
-            Supplier<Map<ColumnHandle, NullableValue>> partitionValues = memoize(() -> getPartitionValues(identityPartitionColumns, partitionKeys));
+                Supplier<Map<ColumnHandle, NullableValue>> partitionValues = memoize(() -> getPartitionValues(identityPartitionColumns, partitionKeys));
 
-            if (!dynamicFilterPredicate.isAll() && !dynamicFilterPredicate.equals(pushedDownDynamicFilterPredicate)) {
-                if (!partitionMatchesPredicate(
-                        identityPartitionColumns,
-                        partitionValues,
-                        dynamicFilterPredicate)) {
-                    continue;
+                if (!dynamicFilterPredicate.isAll() && !dynamicFilterPredicate.equals(pushedDownDynamicFilterPredicate)) {
+                    if (!partitionMatchesPredicate(
+                            identityPartitionColumns,
+                            partitionValues,
+                            dynamicFilterPredicate)) {
+                        continue;
+                    }
+                    if (!fileMatchesPredicate(
+                            fieldIdToType,
+                            dynamicFilterPredicate,
+                            wholeFileTask.file().lowerBounds(),
+                            wholeFileTask.file().upperBounds(),
+                            wholeFileTask.file().nullValueCounts())) {
+                        continue;
+                    }
                 }
-                if (!fileMatchesPredicate(
-                        fieldIdToType,
-                        dynamicFilterPredicate,
-                        scanTask.file().lowerBounds(),
-                        scanTask.file().upperBounds(),
-                        scanTask.file().nullValueCounts())) {
+                if (!partitionMatchesConstraint(identityPartitionColumns, partitionValues, constraint)) {
                     continue;
                 }
```

Comment:
Could you pls extract all these checks (yes, they only check whether the file scan task should be pruned) to a separate method?
Doing this would keep the logic within `while` much easier to follow.